### PR TITLE
Permissions pr3 part2

### PR DESCRIPTION
**Pull request by dantrinh**
## Description

1. Added gating on view chat channels, view chat messages, and the chat sidebar
2. Created new migration to replace chat_enabled with default_allow_permissions and default_deny_permissions
3. added these new allow and deny permission fields into Chain Model and ChainInfo
4. Updated bulkoffchain to return community roles per chainid and store them in chaininfo
5. wrote needed helpers for checking active address permissiosn and etc in controllers/server/roles.ts and also the util roles file. 

## How has this been tested?

1. Did QA on checking the chat sidebar, but was not able to check the view chat channels and view chat messages

## Does this PR affect any server routes?
- [ ] yes
- [ ] no

## Have proper tags been added (for bug, enhancement, breaking change)?
- [ ] yes

### CODE REVIEW

File: packages/commonwealth/client/scripts/controllers/server/roles.ts [hunk 1]
```diff
@@ -299,3 +307,61 @@ export class RolesController {
...
+  chain_info: ChainInfo,
+  action: Action
+): PermissionError | undefined {
+  const chainRoles = active_address_roles.filter(
+    (r) => r.chain_id === chain_info.id
+  );
+
+  if (chainRoles.length > 0) {
+    // sort roles by roles with lowest roles first
+    chainRoles.sort((a) => {
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/controllers/server/roles.ts:323]: All this logic is basically replicated -- any way we can make it into a helper, for consistency?

File: packages/commonwealth/client/scripts/views/components/sidebar/index.tsx [hunk 2]
```diff
@@ -23,9 +30,18 @@ type SidebarAttrs = {
...
-    const hideChat = !app.chain?.meta?.chatEnabled;
-
+    const activeAddressRoles = app.roles.getAllRolesInCommunity({
+      chain: app.activeChainId(),
+    });
+    const currentChainInfo = app.chain?.meta;
+    const hideChat =
+      currentChainInfo &&
+      activeAddressRoles &&
+      isActiveAddressPermitted(
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/views/components/sidebar/index.tsx:40]: This doesn't feel clean to me -- makes more sense on server. On client, can we have it return a boolean instead?

File: packages/commonwealth/client/scripts/views/pages/manage_community/chain_metadata_rows.tsx [hunk 3]
```diff
@@ -23,6 +23,12 @@ import {
   setChainCategories,
 } from './helpers';
 import { CWLabel } from '../../components/component_kit/cw_label';
+import {
+  Action,
+  addPermission,
+  isPermitted,
+  removePermission,
+} from '../../../../../../common-common/src/permissions';
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/views/pages/manage_community/chain_metadata_rows.tsx:31]: You can reference `common-common` directly, no need for relative import.

File: packages/commonwealth/client/scripts/views/pages/manage_community/helpers.ts [hunk 4]
```diff
@@ -40,9 +40,13 @@ export const setChainCategories = async (
             },
           ]);
         } else if (!create) {
-          app.config.chainCategories = app.config.chainCategories.filter(
-            (data) => data.id !== response.result.id
-          );
+          // TODO: this is a hack, we should be able to just remove the category from the list
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/views/pages/manage_community/helpers.ts:43]: Can you explain what this does / affects? I don't understand.

File: packages/commonwealth/server/routes/chat/getChatChannel.ts [hunk 5]
```diff
@@ -14,17 +21,65 @@ export const Errors = {
...
+  // check channel id
+  if (!req.query.channel_id) {
+    return next(new AppError(Errors.NoChannelId));
+  }
+  //   check chain id
+  if (!req.query.chain_id) {
+    return next(new AppError(Errors.NoChainId));
+  }
+  //   If not logged in check if the channel is public
+  if (!req.user?.id) {
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/chat/getChatChannel.ts:39]: We need to be checking based on whether or not `getActiveAddress` returns something or nothing -- i.e. if there IS an activeAddress, we use `isAddressPermitted`. Otherwise, we use `isAnyonePermitted`.

This entire pattern should be abstracted away into a single check.

File: packages/commonwealth/server/routes/chat/getChatMessages.ts [hunk 6]
```diff
@@ -1,52 +1,100 @@
...
-
-	return res.json({ status: '200', result: JSON.stringify(messages) });
-}
+export default async (
+  models: DB,
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  if (!req.user) {
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/chat/getChatMessages.ts:31]: See comments on other route -- we should be abstracting this duplicated code into another helper fn.

File: packages/commonwealth/server/routes/updateChain.ts [hunk 7]
```diff
@@ -133,8 +135,12 @@ const updateChain = async (
   if (custom_stages) chain.custom_stages = custom_stages;
   if (terms) chain.terms = terms;
   if (snapshot) chain.snapshot = snapshot;
-  if (chat_enabled) chain.chat_enabled = chat_enabled;
+  // Set default allow/deny permissions
+  chain.default_allow_permissions = default_allow_permissions;
+  chain.default_deny_permissions = default_deny_permissions;
 
+  chain.default_allow_permissions = default_allow_permissions || BigInt(0);
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/updateChain.ts:142]: Why do we set them twice and do a meaningless `|| BigInt(0)`?

File: packages/commonwealth/test/unit/permissions/isActiveAddressPermitted.spec.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
+// import { assert, expect } from 'chai';
```
Comment (jnaviask) [inline packages/commonwealth/test/unit/permissions/isActiveAddressPermitted.spec.ts:1]: What's going on here?

File: packages/commonwealth/test/unit/rules/rolesHelpers.spec.ts [hunk 9]
```diff
@@ -24,34 +24,7 @@ describe('getHighestRole() unit tests', () => {
         deny: BigInt(0),
       },
     ];
-    const role = await getHighestRole(roles);
+    const role = await getHighestRoleFromCommunityRoles(roles);
     assert.equal(role.name, 'admin');
   });
 });
-
-describe('getHighestRole() unit tests', () => {
```
Comment (jnaviask) [inline packages/commonwealth/test/unit/rules/rolesHelpers.spec.ts:32]: Why did you remove the unit test?

**Review by jnaviask — CHANGES_REQUESTED** [review]

File: packages/commonwealth/client/scripts/views/components/sidebar/index.tsx (hunk 2 above)
Comment (dantrinh) [reply packages/commonwealth/client/scripts/views/components/sidebar/index.tsx:40]: yes!

File: packages/commonwealth/client/scripts/views/pages/manage_community/chain_metadata_rows.tsx (hunk 3 above)
Comment (dantrinh) [reply packages/commonwealth/client/scripts/views/pages/manage_community/chain_metadata_rows.tsx:31]: yes!

File: packages/commonwealth/client/scripts/views/pages/manage_community/helpers.ts (hunk 4 above)
Comment (dantrinh) [reply packages/commonwealth/client/scripts/views/pages/manage_community/helpers.ts:43]: There are times where there is no response returned and it is null causing the chainEntity routes to error. This only performs the chainCategory removal if the response is nonNull

File: packages/commonwealth/test/unit/rules/rolesHelpers.spec.ts (hunk 9 above)
Comment (dantrinh) [reply packages/commonwealth/test/unit/rules/rolesHelpers.spec.ts:32]: it was a duplicate test, I accidentally copied it

File: packages/commonwealth/server/routes/chat/getChatChannel.ts (hunk 5 above)
Comment (dantrinh) [reply packages/commonwealth/server/routes/chat/getChatChannel.ts:38]: put this in a helper in the server utils!

File: packages/commonwealth/client/scripts/controllers/server/roles.ts (hunk 1 above)
Comment (dantrinh) [reply packages/commonwealth/client/scripts/controllers/server/roles.ts:323]: I made a helper and stored in shared/utils so that the client and server can use it

File: packages/commonwealth/server/routes/updateChain.ts (hunk 7 above)
Comment (dantrinh) [reply packages/commonwealth/server/routes/updateChain.ts:139]: the || BigInt(0) is to guarantee that there is a default value set if default_allow_permissions is null. The previous lines are useless though I just removed them

**Comment by dantrinh** [comment]
@jnaviask fixed a lot of internal bugs with the permission helper functions and added better testing. Should be ready for rereview!

**Comment by dantrinh** [comment]
@jnaviask Ready freddy! + New Tests and fixed missed edge cases

File: packages/commonwealth/client/scripts/controllers/server/roles.ts [hunk 10]
```diff
@@ -8,6 +8,15 @@ import {
...
 } from 'models';
+import {
+  Action,
+  BASE_PERMISSIONS,
+  computePermissions,
+  isPermitted,
+  PermissionError,
+  Permissions,
+} from 'common-common/src/permissions';
+import { aggregatePermissions } from 'commonwealth/shared/utils';
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/controllers/server/roles.ts:19]: Remove commonwealth prefix.

File: packages/commonwealth/client/scripts/views/components/sidebar/index.tsx [hunk 11]
```diff
@@ -23,9 +30,18 @@ type SidebarAttrs = {
...
   view(vnode: m.VnodeDOM<SidebarAttrs, this>) {
     const { isSidebarToggleable, isSidebarToggled } = vnode.attrs;
-
-    const hideChat = !app.chain?.meta?.chatEnabled;
-
+    const activeAddressRoles = app.roles.getAllRolesInCommunity({
+      chain: app.activeChainId(),
+    });
+    const currentChainInfo = app.chain?.meta;
+    const hideChat =
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/views/components/sidebar/index.tsx:37]: `hideChat` should be biased toward hiding rather than showing, e.g. how the older version returned true if chain/meta were undefined, as well as based on the variable.

File: packages/commonwealth/server/routes/bulkOffchain.ts [hunk 12]
```diff
@@ -4,10 +4,10 @@
 // because it's easy to miss catching errors inside the promise executor, but we use it in this file
 // because the bulk offchain queries are heavily optimized so communities can load quickly.
 //
-import { contracts } from '@polkadot/types/interfaces/definitions';
 import { factory, formatFilename } from 'common-common/src/logging';
 import { NextFunction, Request, Response } from 'express';
 import { Op, QueryTypes } from 'sequelize';
+import { CommunityRoleInstance } from 'server/models/community_role';
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/bulkOffchain.ts:10]: Use relative import.

File: packages/commonwealth/server/routes/chat/getChatChannel.ts [hunk 13]
```diff
@@ -14,17 +20,45 @@ export const Errors = {
...
+      models,
+      req.query.chain_id,
+      Action.VIEW_CHAT_CHANNELS
+    );
+    if (permission_error === PermissionError.NOT_PERMITTED) {
+      return next(new AppError(PermissionError.NOT_PERMITTED));
+    }
+  }
+
+  await checkActiveAddressPermitted(
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/chat/getChatChannel.ts:49]: TODO: rewrite these patterns.

File: packages/commonwealth/server/routes/chat/getChatChannel.ts [hunk 14]
```diff
@@ -14,17 +20,45 @@ export const Errors = {
  * @param res
  * @param next
  */
-export default async (models: DB, req: Request, res: Response, next: NextFunction) => {
-	// check channel id
-	if (!req.query.channel_id) {
-		return next(new AppError(Errors.NoChannelId))
-	}
+export default async (
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/chat/getChatChannel.ts:23]: Refactor this into ts route.

File: packages/commonwealth/server/routes/chat/getChatMessages.ts [hunk 15]
```diff
@@ -1,52 +1,72 @@
...
  * @param models
  * @param req
  * @param res
  * @param next
  */
-export default async (models: DB, req: Request, res: Response, next: NextFunction) => {
-	if (!req.user) {
-		return next(new AppError(Errors.NotLoggedIn));
-	}
+export default async (
```
Comment (jnaviask) [inline packages/commonwealth/server/routes/chat/getChatMessages.ts:24]: Refactor into TS route.

**Review by jnaviask — CHANGES_REQUESTED** [review]
Jake to resolve changes.

**Comment by jnaviask** [comment]
Remaining thing to debug is why creating a "private" chat room on Terra doesn't work. Then this is good to go.

**Review by jnaviask — APPROVED** [review]
Some errors with chat that are unrelated to this PR. I think we're good to go.
